sw/device/lib/testing/pwrmgr_testutils.c

```diff
@@ -0,0 +1,37 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/testing/check.h"
+
+void pwrmgr_testutils_enable_low_power(
+    dif_pwrmgr_t *pwrmgr, dif_pwrmgr_request_sources_t wakeups,
+    dif_pwrmgr_domain_config_t domain_config) {
+  // Enable low power on the next WFI with clocks and power domains configured
+  // per domain_config.
+
+  // Issue #6504: USB clock in active power must be left enabled.
+  domain_config |= kDifPwrmgrDomainOptionUsbClockInActivePower;
+
+  CHECK_DIF_OK(
+      dif_pwrmgr_set_request_sources(pwrmgr, kDifPwrmgrReqTypeWakeup, wakeups));
+  CHECK_DIF_OK(dif_pwrmgr_set_domain_config(pwrmgr, domain_config));
+  CHECK_DIF_OK(dif_pwrmgr_low_power_set_enabled(pwrmgr, kDifToggleEnabled));
+}
+
+bool pwrmgr_testutils_is_wakeup_reason(dif_pwrmgr_t *pwrmgr,
+                                       dif_pwrmgr_request_sources_t reasons) {
+  dif_pwrmgr_wakeup_reason_t wakeup_reason;
+  CHECK_DIF_OK(dif_pwrmgr_wakeup_reason_get(pwrmgr, &wakeup_reason));
+
+  return (wakeup_reason.request_sources == 0 ||
+          wakeup_reason.types == kDifPwrmgrWakeupTypeRequest) &&
+         wakeup_reason.request_sources == reasons;
```

Comment:
Maybe I am missing something, but something feels off about this logic. Shouldn't this just be:
```c
  return wakeup_reason.types == kDifPwrmgrWakeupTypeRequest &&
             wakeup_reason.request_sources == reasons;
```

Reply:
Unfortunately not. The implementation is funky: https://github.com/lowRISC/opentitan/blob/45cb5d38bc86d0147801dc3a81130ba29be4ad9c/sw/device/lib/dif/dif_pwrmgr.c#L470

for POR request_sources is 0, so the types field is not set. I don't like this "smart" code... see https://github.com/lowRISC/opentitan/blob/45cb5d38bc86d0147801dc3a81130ba29be4ad9c/sw/device/lib/dif/dif_pwrmgr.c#L120

The "smarts" make it more confusing.

Reply:
Seems like we should fix the DIF then, now that it is intended for our use.

Reply:
Not in this PR, and it seems a low priority item since the testutils hides the gnarly details.